In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
df = pd.read_excel('data_final_res.xlsx')

df.dtypes

Уровень                                                                                                                         int64
Заказчик                                                                                                                       object
ИНН заказчика                                                                                                                   int64
Стоимость(руб.) Заказчик                                                                                                      float64
Реестровый номер публикации                                                                                                     int64
Идентификационный код закупки                                                                                                  object
Сфера деятельности                                                                                                             object
Регион поставки                                               

In [19]:
df.head()

,Уровень,Заказчик,ИНН заказчика,Стоимость(руб.) Заказчик,Реестровый номер публикации,Сфера деятельности,Регион поставки,Город поставки,Дата публикации,Дата окончания приема заявок / Дата планового окончания исполнения контракта / Плановая дата публикации лота по ППГ,...,Дата окончания проведения торгов,Поставщик,ИНН поставщика,"Снижение на торгах,%",Форма публикации,Тип торгов,РНП сейчас,РНП ранее,Target,Допущен
0,2,АДМИНИСТРАЦИЯ ВАРГАШИНСКОГО РАЙОНА,4505003804,11.511685,143300049220000000,[ОКПД2 17.12] Бумага и картон,Курганская область,Варгашинский район,2020-01-20 13:17:13,2020-01-28 10:00:00,...,2020-01-29 00:00:00,Кожина Алла Владимировна,450100192605,0.100001,Торговая процедура,Неизвестно,0,0,1,1
1,2,ГБОУ ШКОЛА № 292 ФРУНЗЕНСКОГО РАЙОНА САНКТ-ПЕТ...,7816165190,10.762961,372200177920000000,[ОКПД2 95.11] Услуги по ремонту компьютеров и ...,Санкт-Петербург,Санкт-Петербург,2020-01-22 17:07:46,2020-01-31 10:00:00,...,2020-02-03 00:00:00,"ООО ""ФОРПРИНТИНГ""",7825449897,0.005024,Торговая процедура,Неизвестно,0,0,0,1
2,1,"МАУК ""ЦБС""",8902012760,13.073226,32211104938,[ОКПД2 35.30] Услуги по снабжению паром и конд...,Ямало-Ненецкий автономный округ,Ямало-Ненецкий автономный округ (Тюменская обл...,2022-02-06 23:57:13,2022-01-31 12:00:00,...,2022-01-31 00:00:00,"АО ""ХАРП-ЭНЕРГО-ГАЗ""",8901016850,0.000000,Торговая процедура,Закупка у единственного поставщика (исполнител...,0,0,1,1
3,2,"ГБУЗ ""ГКБ ИМ. Ф.И. ИНОЗЕМЦЕВА ДЗМ""",7719046800,14.736540,373200052320000000,[ОКПД2 32.50] Инструменты и оборудование медиц...,Москва,Москва,2020-01-15 16:57:11,2020-01-23 09:00:00,...,2020-01-24 00:00:00,Сухачев Александр Анатольевич,343519815501,0.010000,Торговая процедура,Неизвестно,0,0,1,1
4,1,"ГАУ ВО ""МОЙ БИЗНЕС""",3435901976,13.255146,32211612062,[ОКПД2 71.12] Услуги в области инженерно-техни...,Волгоградская область,Волгоградская область,2022-08-10 16:05:14,2022-08-10 11:00:00,...,2022-08-10 11:30:00,"ООО ""ДЕФЕКТАЦИЯ""",3445030277,0.036778,Торговая процедура,Закупка у единственного поставщика (исполнител...,0,0,1,1


In [20]:
df.shape

(109877, 21)

In [25]:
df['Реестровый номер публикации'].unique()

array([143300049220000000, 372200177920000000,        32211104938, ...,
              32413198678,        32211255143,        32008790289],
      shape=(54605,))

In [3]:
date_columns = df.select_dtypes('datetime64[ns]').columns
date_columns

Index(['Дата публикации',
       'Дата окончания приема заявок / Дата планового окончания исполнения контракта / Плановая дата публикации лота по ППГ',
       'Дата начала подачи заявок / Дата начала исполнения контракта / Дата публикации ППГ',
       'Дата окончания проведения торгов'],
      dtype='object')

In [4]:
for col in date_columns:
    df[col] = pd.to_datetime(df[col], format='%Y-%m-%d', errors='coerce')

In [5]:
for col in df.columns:
    print(f"Столбец: {col}")
    print(f"Тип данных: {df[col].dtype}")
    print(f"Количество уникальных значений: {df[col].nunique()}")
    print(f"Количество пропущенных значений: {df[col].isna().sum()}")
    if df[col].nunique() <= 10:
        print(f"Уникальные значения: {df[col].unique()}")
    print("-" * 20)

numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns
for col in numerical_cols:
    print(f'Столбец: {col}')
    print(f'Минимум: {df[col].min()}')
    print(f'Максимум: {df[col].max()}')
    print(f'Среднее: {df[col].mean()}')
    print(f'Медиана: {df[col].median()}')
    print("-" * 20)

Столбец: Уровень
Тип данных: int64
Количество уникальных значений: 2
Количество пропущенных значений: 0
Уникальные значения: [2 1]
--------------------
Столбец: Заказчик
Тип данных: object
Количество уникальных значений: 7669
Количество пропущенных значений: 0
--------------------
Столбец: ИНН заказчика
Тип данных: int64
Количество уникальных значений: 7778
Количество пропущенных значений: 0
--------------------
Столбец: Стоимость(руб.) Заказчик
Тип данных: float64
Количество уникальных значений: 35370
Количество пропущенных значений: 36823
--------------------
Столбец: Реестровый номер публикации
Тип данных: int64
Количество уникальных значений: 54609
Количество пропущенных значений: 0
--------------------
Столбец: Идентификационный код закупки
Тип данных: object
Количество уникальных значений: 12973
Количество пропущенных значений: 76134
--------------------
Столбец: Сфера деятельности
Тип данных: object
Количество уникальных значений: 6595
Количество пропущенных значений: 14
-------

In [6]:
df['Тип торгов'].unique()

array([nan,
       'Закупка у единственного поставщика (исполнителя, подрядчика)',
       'Запрос цен', 'Запрос предложений', 'Прямые закупки',
       'Запрос котировок в электронной форме', 'Конкурс открытый',
       'Аукцион электронный', 'Запрос предложений в электронной форме',
       'Аукцион закрытый', 'Аукцион открытый', 'Иной способ',
       'Запрос предложений с квалификационным отбором',
       'Конкурс открытый с ограниченным участием (с предквалификационным отбором)',
       'Отбор квалификационный', 'Запрос котировок',
       'Маркетинговое исследование',
       'Переговоры конкурентные с квалификационным отбором',
       'Переговоры конкурентные', 'Мелкая закупка',
       'Запрос котировок с квалификационным отбором',
       'Аукцион с квалификационным отбором',
       'Запрос предложений закрытый'], dtype=object)

In [7]:
cross_table = pd.crosstab(df['Победитель'], df['Статус допуска'], margins=True)
print(cross_table)

print("\nРаспределение статуса допуска для победителей:")
print(df[df['Победитель'] == 'Победитель']['Статус допуска'].value_counts(normalize=True))

print("\nРаспределение статуса допуска для не победителей:")
print(df[df['Победитель'].isna()]['Статус допуска'].value_counts(normalize=True))

Статус допуска  Допущен  Не допущен  Неизвестно    All
Победитель                                            
Победитель        62688           2        2834  65524
All               62688           2        2834  65524

Распределение статуса допуска для победителей:
Статус допуска
Допущен       0.956718
Неизвестно    0.043251
Не допущен    0.000031
Name: proportion, dtype: float64

Распределение статуса допуска для не победителей:
Статус допуска
Допущен       0.831744
Неизвестно    0.100976
Не допущен    0.067280
Name: proportion, dtype: float64


Видим, что почти все победители имеют статус "Допущен" (95.7%). Среди не победителей доля "Допущенных" меньше (83.2%), а доли "Неизвестно" и "Не допущен" выше, значит, есть связь между статусом допуска и победой в тендере. Преобразуем "Статус допуска" в бинарный признак, а также заменим "Неизвестно" на "Допущен" только для победителей



In [8]:
df['Target'] = (df['Победитель'] == 'Победитель').astype(int)
df = df.drop('Победитель', axis=1)

df.loc[(df['Target'] == 1) & (df['Статус допуска'] == 'Неизвестно'), 'Статус допуска'] = 'Допущен'

df['Допущен'] = df['Статус допуска'].apply(lambda x: 1 if x == 'Допущен' else 0)
df = df.drop('Статус допуска', axis=1)

In [9]:
for col in ['Стоимость(руб.) Заказчик', 'Идентификационный код закупки', 'Тип торгов']:
    print(f"Столбец: {col}")
    missing_ratio = df[col].isna().sum() / len(df)
    print(f"Доля пропусков: {missing_ratio:.2%}")

   # статистики для строк с пропусками в исследуемом столбце и без пропусков
    if df[col].dtype == 'int64' or df[col].dtype == 'float64':
        print("Статистика для строк с пропусками:")
        print(df[df[col].isnull()].describe())
        print("\nСтатистика для строк без пропусков:")
        print(df[~df[col].isnull()].describe())
    print("-" * 20)

Столбец: Стоимость(руб.) Заказчик
Доля пропусков: 33.51%
Статистика для строк с пропусками:
       Уровень  ИНН заказчика  Стоимость(руб.) Заказчик  \
count  36823.0   3.682300e+04                       0.0   
mean       2.0   5.244888e+09                       NaN   
min        2.0   1.050222e+08                       NaN   
25%        2.0   2.725043e+09                       NaN   
50%        2.0   6.319231e+09                       NaN   
75%        2.0   7.701984e+09                       NaN   
max        2.0   9.909310e+09                       NaN   
std        0.0   2.711643e+09                       NaN   

       Реестровый номер публикации                Дата публикации  \
count                 3.682300e+04                          36823   
mean                  2.846808e+10  2021-11-22 20:26:04.724710400   
min                   5.438300e+04            2019-12-26 00:00:00   
25%                   3.200908e+10            2020-07-29 16:36:47   
50%                   3.211021e

Стоимость: доля пропусков - 33.51%, все пропуски только при уровне 2. Пропуски будем заполнять последовательно - смотрим медиану по тому же заказчику и сфере, если таких данных нет, то заполняем медианой по всему столбцу

Идентификационный код закупки: доля пропусков - 69.28%, очень много, но есть реестровый номер публикации, поэтому можем не рассматривать дальше этот признак

Тип торгов: доля пропусков - 60.23%, распределения по другим столбцам с пропусками и без пропусков отличаются, пропуски могут быть неслучайными, этот признак надо будет учесть в обучении, пропуски заполняем значением "Неизвестно"

Также есть совсем немного пропусков в "Сфере деятелности", можно удалить эти строки

In [10]:
sphere_company_median = df.groupby(['ИНН заказчика', 'Сфера деятельности'])['Стоимость(руб.) Заказчик'].transform('median')
company_median = df.groupby('ИНН заказчика')['Стоимость(руб.) Заказчик'].transform('median')
sphere_median = df.groupby('Сфера деятельности')['Стоимость(руб.) Заказчик'].transform('median')
global_median = df['Стоимость(руб.) Заказчик'].median()

df['Стоимость(руб.) Заказчик'] = np.where(df['Стоимость(руб.) Заказчик'].isna(),
                sphere_company_median.fillna(company_median.fillna(sphere_median.fillna(global_median))),
                df['Стоимость(руб.) Заказчик']
    )  

df = df.drop(['Идентификационный код закупки'], axis=1)

df['Тип торгов'] = df['Тип торгов'].fillna('Неизвестно')

df = df[df['Сфера деятельности'].notna()]

In [11]:
for col in df.columns:
    print(f"Столбец: {col}")
    print(f"Тип данных: {df[col].dtype}")
    print(f"Количество уникальных значений: {df[col].nunique()}")
    print(f"Количество пропущенных значений: {df[col].isna().sum()}")
    if df[col].nunique() <= 10:
        print(f"Уникальные значения: {df[col].unique()}")
    print("-" * 20)

Столбец: Уровень
Тип данных: int64
Количество уникальных значений: 2
Количество пропущенных значений: 0
Уникальные значения: [2 1]
--------------------
Столбец: Заказчик
Тип данных: object
Количество уникальных значений: 7668
Количество пропущенных значений: 0
--------------------
Столбец: ИНН заказчика
Тип данных: int64
Количество уникальных значений: 7777
Количество пропущенных значений: 0
--------------------
Столбец: Стоимость(руб.) Заказчик
Тип данных: float64
Количество уникальных значений: 36074
Количество пропущенных значений: 0
--------------------
Столбец: Реестровый номер публикации
Тип данных: int64
Количество уникальных значений: 54605
Количество пропущенных значений: 0
--------------------
Столбец: Сфера деятельности
Тип данных: object
Количество уникальных значений: 6595
Количество пропущенных значений: 0
--------------------
Столбец: Регион поставки
Тип данных: object
Количество уникальных значений: 130
Количество пропущенных значений: 0
--------------------
Столбец: Го

In [12]:
print(df['Стоимость(руб.) Заказчик'].describe())

count    1.098770e+05
mean     6.823351e+06
std      1.498381e+08
min      2.000000e+00
25%      2.233226e+05
50%      5.161660e+05
75%      1.806000e+06
max      3.900694e+10
Name: Стоимость(руб.) Заказчик, dtype: float64


In [13]:
with pd.option_context('display.float_format', '{:.2f}'.format):
    print(df['Стоимость(руб.) Заказчик'].describe())

print("\nМедиана:", df['Стоимость(руб.) Заказчик'].median())
print("90% перцентиль:", df['Стоимость(руб.) Заказчик'].quantile(0.9))
print("99% перцентиль:", df['Стоимость(руб.) Заказчик'].quantile(0.99))


count        109877.00
mean        6823350.53
std       149838078.97
min               2.00
25%          223322.58
50%          516166.00
75%         1806000.00
max     39006943870.00
Name: Стоимость(руб.) Заказчик, dtype: float64

Медиана: 516166.0
90% перцентиль: 6208236.0
99% перцентиль: 61316462.0


Судя по данным есть длинный хвост в распределении показателя стоимости, то есть есть небольшое количество очень больших значений, которые сильно влияют на среднее значение, а это может потом подпортить результаты моделей, поэтому логарифмируем столбец, так влияние выбросов будет меньше

In [14]:
df['Стоимость(руб.) Заказчик'] = np.log1p(df['Стоимость(руб.) Заказчик'])

with pd.option_context('display.float_format', '{:.2f}'.format):
    print(df['Стоимость(руб.) Заказчик'].describe())

count   109877.00
mean        13.40
std          1.69
min          1.10
25%         12.32
50%         13.15
75%         14.41
max         24.39
Name: Стоимость(руб.) Заказчик, dtype: float64


In [15]:
print("Уникальные значения в столбце 'Регион поставки':")
print(df['Регион поставки'].unique())

Уникальные значения в столбце 'Регион поставки':
['Курганская область' 'Санкт-Петербург'
 'Ямало-Ненецкий автономный округ (Тюменская область)' 'Москва'
 'Волгоградская область' 'Ростовская область' 'Самарская область'
 'Алтайский край' 'Омская область' 'Республика Башкортостан'
 'Вологодская область' 'Брянская область' 'Республика Бурятия'
 'Республика Крым' 'Владимирская область' 'Саратовская область'
 'Тюменская область' 'Республика Марий Эл' 'Челябинская область'
 'Республика Татарстан (Татарстан)'
 'Ханты-Мансийский автономный округ - Югра (Тюменская область)'
 'Кемеровская область - Кузбасс' 'Ленинградская область'
 'Архангельская область' 'Калужская область' 'Тамбовская область'
 'Хабаровский край' 'Ставропольский край' 'Краснодарский край'
 'Ярославская область' 'Кировская область' 'Республика Коми'
 'Калининградская область' 'Красноярский край' 'Республика Саха (Якутия)'
 'Приморский край' 'Костромская область' 'Нижегородская область'
 'Амурская область' 'Тверская область' 'Но

In [16]:
def clean_region_data(region):
    if pd.isna(region):
        return region
        
    if ';' in region:
        region = region.split(';')[0].strip()
    
    if '(' in region:
        region = region.split('(')[0].strip()
        
    return region

df['Регион поставки'] = df['Регион поставки'].apply(clean_region_data)

print("Уникальные значения после очистки:")
print(sorted(df['Регион поставки'].unique()))

Уникальные значения после очистки:
['Алтайский край', 'Амурская область', 'Архангельская область', 'Астраханская область', 'Байконур', 'Белгородская область', 'Брянская область', 'Владимирская область', 'Волгоградская область', 'Вологодская область', 'Воронежская область', 'Донецкая Народная Республика', 'Еврейская автономная область', 'Забайкальский край', 'Ивановская область', 'Иркутская область', 'Кабардино-Балкарская Республика', 'Калининградская область', 'Калужская область', 'Камчатский край', 'Карачаево-Черкесская Республика', 'Кемеровская область - Кузбасс', 'Кировская область', 'Костромская область', 'Краснодарский край', 'Красноярский край', 'Курганская область', 'Курская область', 'Ленинградская область', 'Липецкая область', 'Магаданская область', 'Москва', 'Московская область', 'Мурманская область', 'Ненецкий автономный округ', 'Нижегородская область', 'Новгородская область', 'Новосибирская область', 'Омская область', 'Оренбургская область', 'Орловская область', 'Пензенская

На основе всего анализа выше можно написать следующую функцию для предобработки данных:

In [17]:
def eda(df):
    
    date_columns = df.select_dtypes('datetime64[ns]').columns
    for col in date_columns:
        df[col] = pd.to_datetime(df[col], format='%Y-%m-%d', errors='coerce')
    
    df['Target'] = (df['Победитель'] == 'Победитель').astype(int)
    df = df.drop('Победитель', axis=1)

    df.loc[(df['Target'] == 1) & (df['Статус допуска'] == 'Неизвестно'), 'Статус допуска'] = 'Допущен'

    df['Допущен'] = df['Статус допуска'].apply(lambda x: 1 if x == 'Допущен' else 0)
    df = df.drop('Статус допуска', axis=1)

    sphere_company_median = df.groupby(['ИНН заказчика', 'Сфера деятельности'])['Стоимость(руб.) Заказчик'].transform('median')
    company_median = df.groupby('ИНН заказчика')['Стоимость(руб.) Заказчик'].transform('median')
    sphere_median = df.groupby('Сфера деятельности')['Стоимость(руб.) Заказчик'].transform('median')
    global_median = df['Стоимость(руб.) Заказчик'].median()

    df['Стоимость(руб.) Заказчик'] = np.where(df['Стоимость(руб.) Заказчик'].isna(),
                    sphere_company_median.fillna(company_median.fillna(sphere_median.fillna(global_median))),
                    df["Стоимость(руб.) Заказчик"]
        )

    df = df.drop(['Идентификационный код закупки'], axis=1)

    df['Тип торгов'] = df['Тип торгов'].fillna('Неизвестно')

    df = df[df['Сфера деятельности'].notna()]

    df['Стоимость(руб.) Заказчик'] = np.log1p(df['Стоимость(руб.) Заказчик'])

    df['Регион поставки'] = df['Регион поставки'].apply(clean_region_data)

    return df